### ISSUE TITLE
Issue with akka scheduler and poisonpill in UTC

### ISSUE THREAD

**Issue by samikrc** [akka/akka#21787]
Hi,

We have written an akka actor class (SessionEventProcessor) which is called by its parent actor class “EventAndResultProcessor”. EventAndResultProcessor creates a different instance of this actor based on a key called “channelSessionId”. This actor keeps running till it processes all the events (for that particular “channelSessionId”) that comes into it. Once all the events are processed, the actor waits for an additional 180 sec (for more events) and sends poison pill to kill itself (“SessionEventProcessor-channelSessionId” actor).

Here is the relevant part of the code:
```
case processingResult: ProcessingResultMessage =>
{
	...

	if (!eventQueue.isEmpty)
	{
		processEvent(eventQueue.dequeue)
	}
	else
	{
		...
		log.debug(s"Done processing all events for: $channelSessionId.")
		terminateMessage = context.system.scheduler.scheduleOnce(180 seconds, self, new TerminateMessage)
	}
}
case _: TerminateMessage =>
{
	log.debug(s"Killing: ${self.path.name.toString}")
	// Schedule a PoisonPill
	context.system.scheduler.scheduleOnce(1 microseconds, self, PoisonPill)
}
```

We are facing an issue while running this program.

When we run this program some of the actors (SessionEventProcessor) started by the EventAndResultProcessor are getting killed early. In these cases, the server time has not reached 180 secs, but the UTC time suddenly adds 180 sec to itself. I am assuming that the akka scheduler looks at the UTC time and decides to kill the actor (by sending the poisonpill to itself).

Please see the below logs taken while running the program:

**2016-10-28 16:36:11,180 11:03:01.274UTC** DEBUG [EventProcessorApp-akka.actor.default-dispatcher-2] c.i.d.s.SessionEventsProcessor        ## Sending WebSessionStartEvent t
o Actor[akka://EventProcessorApp/user/eventAndResultProcessor/sessionEventsProcessor-05059904-e5d3-4cca-826f-4699ad6e7dbd::4-951/staticProcessor-#560961127] 05059904-e5d
3-4cca-826f-4699ad6e7dbd::4
**2016-10-28 16:36:11,181 11:03:01.283UTC** DEBUG [EventProcessorApp-akka.actor.default-dispatcher-2] c.i.d.s.SessionEventsProcessor Writing to graph State [GraphElementProp
erty: node/session/channelSessionId/05059904-e5d3-4cca-826f-4699ad6e7dbd::4/visitorid/05059904-e5d3-4cca-826f-4699ad6e7dbd]
**2016-10-28 16:36:11,181 11:03:01.283UTC** DEBUG [EventProcessorApp-akka.actor.default-dispatcher-2] c.i.d.s.SessionEventsProcessor Writing to graph State [GraphElementProp
erty: node/session/channelSessionId/05059904-e5d3-4cca-826f-4699ad6e7dbd::4/rv/77.561]
**2016-10-28 16:36:11,181 11:03:01.283UTC** DEBUG [EventProcessorApp-akka.actor.default-dispatcher-2] c.i.d.s.SessionEventsProcessor Writing to graph State [GraphElementProp
erty: node/session/channelSessionId/05059904-e5d3-4cca-826f-4699ad6e7dbd::4/referrer/xxx]
**2016-10-28 16:36:11,181 11:03:01.283UTC** DEBUG [EventProcessorApp-akka.actor.default-dispatcher-2] c.i.d.s.SessionEventsProcessor From akka://EventProcessorApp/user/event
AndResultProcessor/sessionEventsProcessor-05059904-e5d3-4cca-826f-4699ad6e7dbd::4-951/staticProcessor-: 05059904-e5d3-4cca-826f-4699ad6e7dbd::4: events left = 0, process
ing status: SeenNoEndEvents
**2016-10-28 16:36:11,181 11:03:01.283UTC** DEBUG [EventProcessorApp-akka.actor.default-dispatcher-2] c.i.d.s.SessionEventsProcessor Done processing all events for: 05059904
-e5d3-4cca-826f-4699ad6e7dbd::4.
**2016-10-28 16:36:16,440 _11:06:01.304UTC_** DEBUG [EventProcessorApp-akka.actor.default-dispatcher-6] c.i.d.s.SessionEventsProcessor Killing: sessionEventsProcessor-05059904-e5d3-4cca-826f-4699ad6e7dbd::4-951
**2016-10-28 16:36:16,444 _11:06:01.333UTC_** DEBUG [EventProcessorApp-akka.actor.default-dispatcher-5] c.i.d.s.SessionEventsProcessor Stopping: sessionEventsProcessor-05059904-e5d3-4cca-826f-4699ad6e7dbd::4-951

If you notice the server time (which is in IST), it is increasing and the UTC timings are also increasing the same way. But suddenly the UTC timing increases to “11:06:01.304UTC” (in bold-italics) from “11:03:01.283UTC” whereas the server time has increased only by 5 sec (“2016-10-28 16:36:16,440” from “2016-10-28 16:36:11,181”). 

Has anyone else faced this issue, or are we doing something wrong? Kindly let me know if more information is needed. We are using Scala 2.11.8 and akka 2.4.12 on Java 8.

Thanks
-Samik

**Comment by johanandren** [comment]
Hi Samik,

The Akka scheduler uses `System.nanotime` for relative time calculations, which is not based on wall time so your timezone log discrepancies are likely not related at all.

I would guess your actor that schedules the termination message earlier than you expect, if you are sur it is not, could you please see if you could create a complete minimal repeater that you can share with us?

**Comment by drewhk** [comment]
> The Akka scheduler uses System.nanotime for relative time calculations, which is not based on wall time so your timezone log discrepancies are likely not related at all.

It depends. Is this a virtualized environment? In some cases virtual servers *might* change even the HPC counters you see, i.e. breaking nanoTime contracts. Properly configured, it should not do so, but I have *seen things*.

**Comment by tpgirishn** [comment]
2016-10-28 16:36:11,181 11:03:01.283UTC DEBUG [EventProcessorApp-akka.actor.default-**_dispatcher-2_**] c.i.d.s.SessionEventsProcessor Done processing all events for: 05059904
-e5d3-4cca-826f-4699ad6e7dbd::4.

2016-10-28 16:36:16,440 11:06:01.304UTC DEBUG [EventProcessorApp-akka.actor.default-_**dispatcher-6**_] c.i.d.s.SessionEventsProcessor Killing: sessionEventsProcessor-05059904-e5d3-4cca-826f-4699ad6e7dbd::4-951

The dispatchers are different I guess.

**Comment by drewhk** [comment]
Those are just different threads of the default-dispatcher.

**Comment by samikrc** [comment]
Forgot to add comments to this thread and update. We are still grappling with this issue, and in fact, this is one of the reasons we have decided to move our stream processing to Flink, which interestingly, uses akka under the hood. However, we do not have enough time to debug this any further.

The problem shows up in both VMs and actual systems. I tried extracting out the part of the code so that I can reproduce with minimal code, but have not been able to complete it for lack of time.

@tpgirishn : What were you indicating when you pointed out that the dispatchers are different? I would think that irrespective of the dispatchers, the time would flow at the same rate in UTC clock and wall clock.

Thanks.
